core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -166,7 +166,12 @@ inline void configure_shmem_preference(KernelFuncPtr const& func,
         (prefer_shmem ? cudaFuncCachePreferShared : cudaFuncCachePreferL1)));
     return prefer_shmem;
   }();
-  (void)cache_config_preference_cached;
+  if (cache_config_preference_cached != prefer_shmem) {
+    CUDA_SAFE_CALL(cudaFuncSetCacheConfig(
+        func,
+        (prefer_shmem ? cudaFuncCachePreferShared : cudaFuncCachePreferL1)));
+    cache_config_preference_cached = prefer_shmem;
```

Comment:
You should be able to avoid repeating the lambda by just saving it, i.e.
```C++
auto set_preferred_cache_lambda = [&] { ... };
static bool cache_config_preference_cached = set_preferred_cache_lambda();
if (cache_config_preference_cached != prefer_shmem)
  cache_config_preference_cached = set_preferred_cache_lambda();
```

Reply:
Done